### PR TITLE
[TRTLLM-11228][feat] Support DFlash in one-model spec dec

### PR DESCRIPTION
**Pull request by ziyixiong-nv**
## New Features
* Added DFlash speculative decoding mode as a new speculative decoding strategy.
* Added DFlashDecodingConfig enabling configuration of mask token ID and target layer IDs for draft model parameters.
* Integrated DFlash worker and metadata implementations into the speculative decoding pipeline.

## Tests
* Added unit tests for DFlash functionality.
* Added integration accuracy tests with Llama 3.1 8B, Qwen3 8B, and GPT-OSS 20B models.

## Description

The PR is to support DFlash (https://arxiv.org/pdf/2602.06036 ) in one-model spec dec.

In the mtbench testiing on Qwen3-8B-FP8 on 8xB200, the perf of DFlash is not as good as PARD, and the AR is also lower than PARD.

image

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by xinhe-nv — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_speculative.py [hunk 1]
```diff
@@ -781,6 +782,394 @@ def forward(
...
+        import torch.nn.functional as F
+
+        layer0 = self.model.layers[0]
+        attn0 = layer0.self_attn
+        q_size = attn0.q_size
+        kv_size = attn0.kv_size
+        head_dim = attn0.head_dim
+        num_heads_per_rank = attn0.num_heads
+        num_kv_heads_per_rank = attn0.num_key_value_heads
+        num_heads_per_rank // num_kv_heads_per_rank
```
[the remark this answers is not in the document]
Comment (ziyixiong-nv) [reply tensorrt_llm/_torch/models/modeling_speculative.py]: Updated.

File: tests/unittest/_torch/speculative/test_dflash.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    attn_backend = "TRTLLM"
+    enable_block_reuse = False
+    enable_chunked_prefill = False
+
+    total_mem_gb = torch.cuda.get_device_properties(0).total_memory / 1e9
+    if total_mem_gb < 35:
+        pytest.skip("Not enough memory to load target + draft model")
+
+    models_path = llm_models_root()
+    dflash_model_dir = "/code/llm-models/Qwen3-8B-DFlash-b16"
```
[the remark this answers is not in the document]
Comment (ziyixiong-nv) [reply tests/unittest/_torch/speculative/test_dflash.py]: Updated to use `models_path`.

File: tensorrt_llm/llmapi/llm_args.py [hunk 3]
```diff
@@ -3689,6 +3744,28 @@ def validate_speculative_config(self):
             if isinstance(self.speculative_config, PARDDecodingConfig):
                 assert self.speculative_config.max_draft_len > 0, "PARD max_draft_len must be > 0"
 
+            if isinstance(self.speculative_config, DFlashDecodingConfig):
+                assert self.speculative_config.max_draft_len > 0, "DFlash max_draft_len must be > 0"
+                # Resolve target_layer_ids from draft model config if not set
+                if self.speculative_config.target_layer_ids is None and self.speculative_config.speculative_model is not None:
```
[the remark this answers is not in the document]
Comment (ziyixiong-nv) [reply tensorrt_llm/llmapi/llm_args.py]: Fixed — target_layer_ids and mask_token_id are now resolved independently from the draft config.json.

File: tensorrt_llm/_torch/speculative/dflash.py [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
+        if worker is not None and worker._ctx_buf_inited:
+            current = set(self.request_ids)
+            for rid in list(worker._req_to_slot.keys()):
+                if rid not in current:
+                    slot = worker._req_to_slot.pop(rid)
+                    worker._ctx_len[slot] = 0
+                    worker._free_slots.append(slot)
+
+            mapping = torch.tensor(
+                [worker._req_to_slot.get(rid, 0) for rid in self.request_ids],
```
[the remark this answers is not in the document]
Comment (ziyixiong-nv) [reply tensorrt_llm/_torch/speculative/dflash.py:107]: This does happen during warmup — synthetic requests don't have slot assignments, so erroring out would crash on startup. Added a comment explaining why the default to 0 is intentional.

File: tests/unittest/_torch/speculative/test_dflash.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+from utils.llm_data import llm_models_root
+
+from tensorrt_llm import LLM, SamplingParams
+from tensorrt_llm.llmapi import CudaGraphConfig, DFlashDecodingConfig, KvCacheConfig
+
+sys.path.append(os.path.join(os.path.dirname(__file__), ".."))
+
+
+@pytest.fixture(scope="function")
+def enforce_single_worker(monkeypatch):
```
[the remark this answers is not in the document]
Comment (ziyixiong-nv) [reply tests/unittest/_torch/speculative/test_dflash.py]: Removed as it's not used.

**Review by mikeiovine — APPROVED** [review]

File: tests/unittest/_torch/speculative/test_dflash.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        attn_backend=attn_backend,
+        disable_overlap_scheduler=disable_overlap_scheduler,
+        cuda_graph_config=cuda_graph_config,
+        max_batch_size=max_batch_size,
+        kv_cache_config=kv_cache_config,
+        max_seq_len=2048,
+        enable_chunked_prefill=enable_chunked_prefill,
+    )
+
+    spec_config = DFlashDecodingConfig(
```
Comment (nv-guomingz) [inline tests/unittest/_torch/speculative/test_dflash.py:54]: Could we also add DFlash into both quickstart_advanced.py https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/llm-api/quickstart_advanced.py#L304 and https://github.com/NVIDIA/TensorRT-LLM/blob/main/docs/source/features/speculative-decoding.md ?

**Review by nv-guomingz — APPROVED** [review]

**Review by lucaslie — APPROVED** [review]
API changes look good to me

File: tests/unittest/_torch/speculative/test_dflash.py (hunk 6 above)
Comment (ziyixiong-nv) [reply tests/unittest/_torch/speculative/test_dflash.py:54]: Thanks for catching this. I'l do it in a separate PR once this one got merged.

**Review by arysef — APPROVED** [review]
